### ISSUE TITLE
loss non - convergence in CopulaGAN

### ISSUE THREAD

**Issue by HarisNaveed17** [sdv-dev/SDV#336]
### Environment details

If you are already running SDV, please indicate the following details about the environment in
which you are running it:

* SDV version: 0.8.0
* Python version: 3.8.5
* Operating System: Ubuntu 18.04

### Problem description

I'm working with a data set of 7700 examples and 7 features. All features have values greater than zero. Four of these columns are float values and three are integers. If I train the CopulaGAN on it directly without changing any hyper-parameters, both the generator loss (loss G) and discriminator loss (loss D) seem to constantly increase/explode as the epochs progress. 

Epoch 1, Loss G: -0.3189,Loss D: -0.0465
Epoch 2, Loss G: -0.7920,Loss D: -1.7224
Epoch 3, Loss G: -1.1360,Loss D: -1.4083
Epoch 4, Loss G: -1.2272,Loss D: -0.9662
Epoch 5, Loss G: -1.3408,Loss D: -1.0049
Epoch 6, Loss G: -0.9360,Loss D: -0.9909
Epoch 7, Loss G: -0.4682,Loss D: -0.8540
Epoch 8, Loss G: -0.5000,Loss D: -0.9931
Epoch 9, Loss G:  0.0906,Loss D: -0.9765
Epoch 10, Loss G:  0.9288,Loss D: -1.2409
......
Epoch 150, Loss G:  9.8107,Loss D: -164.7765
Epoch 151, Loss G:  10.1032,Loss D: -177.8894
Epoch 152, Loss G:  10.0260,Loss D: -170.7742
Epoch 153, Loss G:  9.9374,Loss D: -168.6565
Epoch 154, Loss G:  10.1594,Loss D: -175.2986
Epoch 155, Loss G:  10.2431,Loss D: -180.0276
Epoch 156, Loss G:  10.5372,Loss D: -184.5792
Epoch 157, Loss G:  10.3357,Loss D: -182.4831
Epoch 158, Loss G:  10.2027,Loss D: -182.7283
Epoch 159, Loss G:  10.3827,Loss D: -187.4515
Epoch 160, Loss G:  10.6095,Loss D: -191.9565
.......
Epoch 290, Loss G:  16.1899,Loss D: -534.5078
Epoch 291, Loss G:  16.5296,Loss D: -523.3918
Epoch 292, Loss G:  16.0048,Loss D: -539.1217
Epoch 293, Loss G:  16.3696,Loss D: -539.8887
Epoch 294, Loss G:  16.5515,Loss D: -532.5117
Epoch 295, Loss G:  16.0033,Loss D: -534.9286
Epoch 296, Loss G:  16.7724,Loss D: -558.3572
Epoch 297, Loss G:  16.3424,Loss D: -557.7731
Epoch 298, Loss G:  16.7596,Loss D: -550.4157
Epoch 299, Loss G:  16.3776,Loss D: -553.1183
Epoch 300, Loss G:  17.2308,Loss D: -569.9644

### What I already tried

I set the field_distributions for each column to gamma, then specified field transformers for the integer columns in the data. I also played around with the embedding_dim, the number of epochs and other parameters. This improved the evaluation score from 0.215 to 0.319, which isn't particularly significant, but the losses increase in the same way.
The same pattern can be seen when using the simple CTGAN as well. However, using the standalone CTGAN package the losses converge to 0 very quickly as shown here:

Epoch 1, Loss G:  2.5657,Loss D: -0.0943
Epoch 2, Loss G:  2.1840,Loss D: -0.0314
Epoch 3, Loss G:  2.0611,Loss D: -0.0463
Epoch 4, Loss G:  2.1895,Loss D: -0.1879
Epoch 5, Loss G:  1.9988,Loss D:  0.0203
Epoch 6, Loss G:  1.8962,Loss D:  0.0326
Epoch 7, Loss G:  1.7505,Loss D: -0.0281
Epoch 8, Loss G:  1.6259,Loss D:  0.0287
Epoch 9, Loss G:  1.4901,Loss D:  0.0213
Epoch 10, Loss G:  1.4011,Loss D: -0.0027
.....
Epoch 290, Loss G: -0.4755,Loss D:  0.0363
Epoch 291, Loss G: -0.5001,Loss D:  0.0172
Epoch 292, Loss G: -0.5793,Loss D:  0.0423
Epoch 293, Loss G: -0.5184,Loss D:  0.0252
Epoch 294, Loss G: -0.4264,Loss D: -0.1161
Epoch 295, Loss G: -0.3500,Loss D:  0.0166
Epoch 296, Loss G: -0.3784,Loss D: -0.1623
Epoch 297, Loss G: -0.3719,Loss D: -0.2994
Epoch 298, Loss G: -0.5721,Loss D:  0.0377
Epoch 299, Loss G: -0.5360,Loss D: -0.1139
Epoch 300, Loss G: -0.4000,Loss D:  0.0140

However, using the evaluate function after sampling from the trained CopulaGAN and standalone CTGAN results in 0.215.
Is there a particular reason why this is so? What am I missing here? From what I understood of the source code the CopulaGAN and CTGAN should train the same since the CopulaGAN calls the CTGAN for training.

**Comment by csala** [comment]
When working on an adversarial setup like the one used on CTGAN, the exact loss values of each component are very often irrelevant. In CTGAN in particular, you should only be concerned if both the generator and the discriminator losses increased at the same time, which is not the case. Here the G loss goes up but the D loss keeps going down faster. This probably means that D is learning faster than G and as a result the overall ability of G to fool D decreases. This may be a bit different than what is usually expected but it does not mean that G stopped learning at any point.

> I set the field_distributions for each column to gamma, then specified field transformers for the integer columns in the data. I also played around with the embedding_dim, the number of epochs and other parameters. This improved the evaluation score from 0.215 to 0.319, which isn't particularly significant, but the losses increase in the same way.

What really matters is this, the evaluation score. And in this case it is indicating that the performance is improving even if the loss values look different that what you expected, so I would not be concerned about it.

Overall, my recommendation here is to focus on getting the highest evaluation score possible and only use the losses as a reference to detect things going wrong if you observe a decrease on the evaluation score.

**Comment by HarisNaveed17** [comment]
Thank you for the explanation. As a quick follow-up, I've also noticed the PAR model's loss begins at different values depending on the segment size and then continuously decreases past zero. I expected that to oscillate around 0 as well. Since PAR is not a GAN - based model and even has a different goal from the tabular models, should I also not worry about its loss values as well?

**Comment by HarisNaveed17** [comment]
Alright. Thanks for the explanation and keep up the good work!
